source/dubregistry/repositories/gitlab.d

```diff
@@ -1,120 +1,121 @@
-			} catch( Exception e ){
-				throw new Exception("Failed to process tag "~tag["name"].get!string~": "~e.msg);
-			}
-		}
-		return ret;
-	}
-
-	RefInfo[] getBranches()
-	{
-		import std.datetime : SysTime;
-
-		Json branches = readJson(getAPIURLPrefix()~"repository/branches?private_token="~m_authToken);
-		RefInfo[] ret;
-		foreach_reverse( branch; branches ){
-			auto branchname = branch["name"].get!string;
-			Json commit = readJson(getAPIURLPrefix()~"repository/commits/"~branch["commit"]["id"].get!string~"?private_token="~m_authToken, true, true);
-			ret ~= RefInfo(branchname, branch["commit"]["id"].get!string, SysTime.fromISOExtString(commit["committed_date"].get!string));
-			logDebug("Found branch for %s/%s: %s", m_owner, m_project, branchname);
-		}
-		return ret;
-	}
-
-	RepositoryInfo getInfo()
-	{
-		RepositoryInfo ret;
-		auto nfo = readJson(getAPIURLPrefix()~"?private_token="~m_authToken);
-		ret.isFork = false; // not reported by API
-		//ret.stars = nfo["star_count"].opt!uint;
-		//ret.forks = nfo["forks_count"].opt!uint;
-		return ret;
-	}
-
-	void readFile(string commit_sha, Path path, scope void delegate(scope InputStream) reader)
-	{
-		assert(path.absolute, "Passed relative path to readFile.");
-		auto url = m_baseURL.toString() ~ (m_owner ~ "/" ~ m_project ~ "/raw/" ~ commit_sha) ~ path.toString() ~ "?private_token="~m_authToken;
-		downloadCached(url, (scope input) {
-			reader(input);
-		}, true);
-	}
-
-	string getDownloadUrl(string ver)
-	{
-		if (m_authToken.length > 0) return null; // don't make private token public
-		if( ver.startsWith("~") ) ver = ver[1 .. $];
-		else ver = ver;
-		return m_baseURL.toString()~m_owner~"/"~m_project~"/repository/archive.zip?ref="~ver;
-	}
-
-	void download(string ver, scope void delegate(scope InputStream) del)
-	{
-		if( ver.startsWith("~") ) ver = ver[1 .. $];
-		else ver = ver;
-		auto url = m_baseURL.toString()~m_owner~"/"~m_project~"/repository/archive.zip?ref="~ver~"&private_token="~m_authToken;
-		downloadCached(url, del);
-	}
-
-	private string getAPIURLPrefix() {
-		return m_baseURL.toString() ~ "api/v3/projects/" ~ (m_owner ~ "/" ~ m_project).urlEncode ~ "/";
-	}
-}
+/**
+	Support for GitLab repositories.
+
+	Copyright: © 2015-2016 rejectedsoftware e.K.
+	License: Subject to the terms of the GNU GPLv3 license, as written in the included LICENSE.txt file.
+	Authors: Sönke Ludwig
+*/
+module dubregistry.repositories.gitlab;
+
+import dubregistry.cache;
+import dubregistry.dbcontroller : DbRepository;
+import dubregistry.repositories.repository;
+import std.string : startsWith;
+import std.typecons;
+import vibe.core.log;
+import vibe.core.stream;
+import vibe.data.json;
+import vibe.inet.url;
+import vibe.textfilter.urlencode;
+
+
+class GitLabRepository : Repository {
+	private {
+		string m_owner;
+		string m_project;
+		URL m_baseURL;
+		string m_authToken;
+	}
+
+	static void register(string auth_token, string url)
+	{
```

Comment:
> By that argument struct and class definitions are also important

This is true to some extent, and I'd probably do it differently there today. The reason it is that way is that the style has been adjusted a few times in the past (mostly to follow language and Phobos style changes).

> anyhow the larger point is "convention" and "conventional" have the same root, which makes an unconventional convention (as this seems to be) stand out oddly. Why not go with something that is so established it even has a name?

This originates from our internal code base. Whatever the style used there is *is the* convention, there is very little reason to care what is used in other groups. And of course I'm using the style that I'm comfortable with for my open source projects, too, especially since they are usually born out of an internal project.

Regarding named "standards", I don't think it makes a big difference in practice if a convention has a name or not. Most people will have to learn the details no matter what, unless everything happens within the same organization/community. I'm all for adapting existing conventions when they are good, but I hate when it's applied dogmatically and gets in the way of better solutions (speaking generally).

Anyway, if you want to change everything DUB to be in Phobos style, I won't object., even if I highly dislike the brace style within function bodies (and I won't start with tabs vs. spaces :wink:).